### PR TITLE
EdDSA support (Ed25519 and Ed448 curves) with Pure, Context and Prehash types + SHA3 (SHA3, SHAKE, CSHAKE and KMAC)

### PR DESCRIPTION
**Pull request by polhenarejos**
## Description
This PR implements EdDSA signatures, with Ed25519 and Ed448 curves. It supports Pure, Context and Prehash types, resulting onto ed25519, ed25519ctx, ed25519ph, ed448 and ed448ph. I followed the RFC 8032, which specifies the procedure for signing and verifying. It also contains verification.

Ed448 uses SHAKE256 function, which is also added. Note that there are no specific CPU optimizations.

I added test suites for EdDSA and SHAKE256 with the test vectors provided in the RFC.

I took the PR from @aurel32, which defined the curves and arithmetic (I fixed some bugs), from #2452.

### About SHA3
During this journey, I started with SHAKE256, since it is necessary for Ed448. However, SHAKE and SHA3 functions are very similar and I extended the sha3 module to include SHA-3, SHAKE, cSHAKE and KMAC (with and without XOF).
Ideally I would make two separate PR, but since EdDSA and SHA-3 are closely dependent I put altogether in this PR.
SHA-3 and SHAKE are based on FIPS 202. cSHAKE and KMAC are based on SP800-185. Test vectors are taken from NIST repository.
**Note: it will only work for architectures that support real qwords (i.e., 64 bits). On smaller CPU Keccak permutation function has to be recompiled with other parameters.**

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
This PR does not contain restarting capability, but it can be extended easily.

Note that this PR obsoletes #3245, since it is contained here and it was not merged.
## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [x] SHA3 module with SHAKE128 and SHAKE256.
- [x] SHA3 with SHA-3 families (224, 256, 384 and 512).
- [x] SHA3 with CSHAKE (128 and 256).
- [x] SHA3 with KMAC (with and without XOF).

## Splitted PR
This PR is being splitted into the following PR:
- #5819 
- #5820
- #5821
- #5822
- #5823

## Related issues
- #2452 
- #3758
- #3757
- #3762
- #3763
- #3245 
- #388

### CODE REVIEW

**Review by gilles-peskine-arm — COMMENTED** [review]
Thank you very much for taking this up!

I've done a first pass of design review. Mainly the `shake256` module/functions will need to be renamed.

A more complete review will unfortunately have to wait since this is a large contribution. We've currently [tentatively scheduled](https://developer.trustedfirmware.org/w/mbed-tls/roadmap/) EdDSA for late 2022, but we may be able to bring it forward, or we may have to delay it further, depending on what else comes in in the meantime.

File: include/mbedtls/mbedtls_config.h [hunk 1]
```diff
@@ -2280,6 +2283,19 @@
...
 
+/**
+ * \def MBEDTLS_EDDSA_C
+ *
+ * Enable the Edwards DSA library.
+ *
+ * Module:  library/eddsa.c
+ * Caller:
+ *
+ * Requires: MBEDTLS_ECP_C, MBEDTLS_ASN1_WRITE_C, MBEDTLS_ASN1_PARSE_C,
```
[the remark this answers is not in the document]
Comment (polhenarejos) [reply include/mbedtls/mbedtls_config.h:2294]: I addressed EdDSA. ECDSA should be addressed in another PR.

**Comment by polhenarejos** [comment]
How can I see the details of failing checks? They point to domain `jenkins-mbedtls.oss.arm.com` but it does not resolv to anywhere.

**Comment by gilles-peskine-arm** [comment]
@polhenarejos The Arm Jenkins instance is not accessible externally, so if you need to see logs, please ask here. We're in the process of moving to OpenCI, which is publicly visible. So far OpenCI is running the same Linux and FreeBSD jobs, but not Windows yet.

On 525ec1fb999c4b777682b36079695c4edd2dafde, builds with MSVC are failing due to this error:

```
C:\builds\workspace\mbed-tls-pr-head_PR-5800-head\src\library\sha3.c(266): error C2220: warning treated as error - no 'object' file generated [C:\builds\workspace\mbed-tls-pr-head_PR-5800-head\src\library\mbedcrypto.vcxproj]
C:\builds\workspace\mbed-tls-pr-head_PR-5800-head\src\library\sha3.c(266): warning C4244: '=' : conversion from 'uint64_t' to 'unsigned char', possible loss of data [C:\builds\workspace\mbed-tls-pr-head_PR-5800-head\src\library\mbedcrypto.vcxproj]
```

**Comment by polhenarejos** [comment]
Thanks @gilles-peskine-arm for the comments and suggestions. PR passes all checks and I think the PR includes all necessary to run EdDSA and SHA3 hashing. Please let me know if it is still missing something or if I can do something else to speed up the merge.

**Comment by gilles-peskine-arm** [comment]
@polhenarejos Thank you very much! And wow, I blinked and now there's the whole SHA3 family? Wow!

Unfortunately our bottleneck is review bandwidth. As this is a very large contribution, we need to schedule it, and I'm afraid we have no room until late 2022.

**Comment by polhenarejos** [comment]
Sure! Once you have shake working is straightforward to add sha3, cshake or kmac. 

In any case, if you think I can help for reviewing to mitigate the bottleneck, tell it to me and I will be glad to give a hand.

**Comment by gilles-peskine-arm** [comment]
Speaking of reviews, would you mind splitting this pull request? It's very large, which makes it difficult to find two reviewers who will have enough time in a reasonably contiguous period. We've never managed to integrate the [previous SHA3 contribution](https://github.com/Mbed-TLS/mbedtls/pull/1549) [originally submitted in 2016](https://github.com/Mbed-TLS/mbedtls/pull/479)... (By the way, did you reuse that for SHA3? Does your code cover the same ground?)

Would you mind splitting this into at least three parts?

1. An EdDSA interface, implemented for Ed25519.
2. SHA3.
3. Extend EdDSA to Ed448.
4. SHAKE (and possibly cSHAKE and KMAC because they are really close).

We'd [planned an even finer breakdown](https://github.com/orgs/Mbed-TLS/projects/1#column-17950143), but since you've already done the work, I think this split would be fine.

**Comment by polhenarejos** [comment]
Honestly, I did not review other contributions for SHA-3 #1549 nor #479, as my intended PR was originally to provide Edwards signatures (and this is why I only added shake256). By a quick overview, I think both do the same but I also add cSHAKE and KMAC (with less code lines). I do not interface each keccak intermediate function because I think this is not useful for anyone and even for the entire keccak-f, which is declared static. 

If it may help, I do no have any objection on splitting in 4 parts. Do you mean submitting as many PR covering each point?
1) and 2) are fine, but 3) requires 4) (Ed448 is defined for shake256 rather than sha3). So, the logic sort would be ed25519+shake256 (and 128)+ed448+sha3+cshake+kmac. Another sort could be ed25519+sha3+shake256/128+edd48+cshake+kmac, depending on how you would prioritize sha3 over ed448 or viceversa. I do not have preferences on that, so whatever you decide is fine to me.

**Comment by gilles-peskine-arm** [comment]
> Do you mean submitting as many PR covering each point?

Yes, please. Or split differently if you think that's more meaningful. I think this is a reasonable split in terms of manageable review size. (That's with code that's already in a good shape. In general we'd ask for even smaller pieces, to reduce the amount of rework if there are a lot of review comments, but I can see that your code is already in a very good shape so I don't expect major difficulties.)

> and 2) are fine, but 3) requires 4)

Yes, sorry, I meant SHA3 then SHAKE then (in either order) Ed448 and the rest of the SHA3 family. Or you could go directly for SHAKE then Ed448 and rest of SHA3 family including SHA3 itself.

**Comment by polhenarejos** [comment]
I added PR:
- #5819 
- #5820 
- #5821 
- #5822 
- #5823 

but I am not sure that they are completely splitted. #5823 adds KMAC support but it also contains the previous commits from the rest of PR.

In addition to this, I do not know how to add a separated PR for Ed448, as it depends on #5819 and #5821. Which is the best approach to do not repeat previous commits?

**Comment by polhenarejos** [comment]
I close this PR as it has been splitted to #5819, #5820, #5821, #5822, #5823 and #5824.

**Pull request closed without merging**